docs/proposals/20210210-insulate-users-from-kubeadm-API-changes.md

```diff
@@ -0,0 +1,260 @@
+
+kubeadm v1beta1 mirror-types:
+
+- Hosted in `bootstrap/kubeadm/types/v1beta1`.
+- Diverged from the original v1beta1 types for better CRD support (in practice
+  a set of `+optional` fixes, few `omitempty` differences).
+
+kubeadm v1beta2 mirror-types:
+
+- Hosted in `bootstrap/kubeadm/types/v1beta2`.
+- Currently, not used in the Cluster API codebase.
+- Does not include changes for better CRD support introduced in kubeadm v1beta1
+  mirror-types.
+
+#### Cluster API v1alpha3 changes
+
+Changes to cluster API v1alpha3 release should be minimal and no breaking change
+should be introduced while implementing this proposal.
+
+__Alternative 1:__
+
+Keep kubeadm v1beta1 types as a Hub type (1); implement conversion to kubeadm API
+version f(Kubernetes Version) when generating the kubeadm config for init/join.
+
+This alternative is the more clean, robust and forward looking, but it requires
+much more work than alternative 2.
+
+__Alternative 2:__
+
+Keep kubeadm v1beta1 types as a Hub type (1); only change the apiVersion to
+`kubeadm.k8s.io/v1beta2` in the generated kubeadm config for init/join.
+
+This alternative is based on following considerations:
+
+- kubeadm v1beta1 mirror types are "compatible" with kubeadm v1beta2 types
+  - Having support for IgnorePreflightErrors into the NodeRegistrationOptions
+    is not required for Cluster API v1alpha3
+  - The automatic certificate copy feature is not used by Cluster API
+  - Improved serialization support has been already applied to the kubeadm
+    v1beta1 mirror-types (and further extended).
+- the minimal Kubernetes version supported by Cluster API is v1.16, and kubeadm
+  v1.16 could work with v1beta2 types. Same for all the other versions up to latest(1.20)
+  and next (1.21).
+- limitations: this approach is not future proof, and it should be reconsidered
+  whenever a new version of kubeadm types is created while v1alpha3 is still supported.
+
+__Common considerations for both alternatives__
+
+KCP is modifying the kubeadm-config Config Map generated by kubeadm, and ideally also
+this bit of code should be made kubeadm version aware.
+
+However, given that the current implementation currently uses unstructured, and
+a change for using the kubeadm types, requires a big refactor, the proposed approach
+for v1alpha3 is to limit the changes to only upgrade the apiVersion when required.
+
+- limitations: this approach is not future proof, and it should be reconsidered
+  whenever a new version of kubeadm types is changing one of the fields
+  edited during upgrades.
+
+(1) See https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts.html
+for a definition of Hub or spoke types/version.
+
+#### Cluster API v1alpha4 changes
+
+Changes to cluster API v1alpha4 could be more invasive and seek for a
+forward-looking solution.
+
+Planned actions are:
+
+- introduce a Cluster API owned version of the kubeadm config types
```

Comment:
would this "version" be unversioned?

Reply:
No, those types will be part of Cluster API types, and thus initially versioned as v1alpha4.
The advantage is that once conversion will be in place, those types are not required anymore to have the same serialization format of the real kubeadm types.

Added a sentence to make this more clear